daemon/service.go

```diff
@@ -0,0 +1,159 @@
+	defer func() {
+		resp.Elapsed = time.Since(start)
+		d.logResponse(resp.Status, req.Language, len(req.Content), resp.Elapsed)
+	}()
+
+	if req.Content == "" {
+		return resp
+	}
+
+	language, dp, err := d.selectPool(req.Language, req.Content, req.Filename)
+	if err != nil {
+		logrus.Errorf("error selecting pool: %s", err)
+		resp.Response = newResponseFromError(err)
+		return resp
+	}
+
+	req.Language = language
+
+	err = dp.Execute(func(driver Driver) error {
+		resp, err = driver.Service().Parse(context.Background(), req)
+		return err
+	})
+
+	if err != nil {
+		resp = &sdk.ParseResponse{}
+		resp.Response = newResponseFromError(err)
+	}
+
+	return resp
+}
+
+func (d *Service) logResponse(s sdk.Status, language string, size int, elapsed time.Duration) {
+	l := logrus.WithFields(logrus.Fields{
+		"language": language,
+		"elapsed":  elapsed,
+	})
+
+	text := fmt.Sprintf("request processed content %d bytes, status %s", size, s)
+
+	switch s {
+	case sdk.Ok:
+		l.Debug(text)
+	case sdk.Error:
+		l.Warning(text)
+	case sdk.Fatal:
+		l.Error(text)
+	}
+}
+
+func (d *Service) NativeParse(req *sdk.NativeParseRequest) *sdk.NativeParseResponse {
+	resp := &sdk.NativeParseResponse{}
+	start := time.Now()
+	defer func() {
+		resp.Elapsed = time.Since(start)
+		d.logResponse(resp.Status, req.Language, len(req.Content), resp.Elapsed)
+	}()
+
+	if req.Content == "" {
+		logrus.Debugf("empty request received, returning empty AST")
+		return resp
+	}
+
+	language, dp, err := d.selectPool(req.Language, req.Content, req.Filename)
+	if err != nil {
+		logrus.Errorf("error selecting pool: %s", err)
+		resp.Response = newResponseFromError(err)
+		return resp
+	}
+
+	req.Language = language
+
+	err = dp.Execute(func(driver Driver) error {
+		resp, err = driver.Service().NativeParse(context.Background(), req)
+		return err
+	})
+
+	if err != nil {
+		resp = &sdk.NativeParseResponse{}
+		resp.Response = newResponseFromError(err)
+	}
+
+	return resp
+}
+
+func (s *Service) selectPool(language, content, filename string) (string, *DriverPool, error) {
+	if language == "" {
+		language = GetLanguage(filename, []byte(content))
+		logrus.Debugf("detected language %q, filename %q", language, filename)
+	}
+
+	dp, err := s.daemon.DriverPool(language)
+	if err != nil {
+		return language, nil, ErrUnexpected.Wrap(err)
+	}
+
+	return language, dp, nil
+}
+
+func (d *Service) Version(req *sdk.VersionRequest) *sdk.VersionResponse {
+	return &sdk.VersionResponse{Version: d.daemon.version}
+}
+
+type ControlService struct {
+	daemon *Daemon
+}
+
+func NewControlService(d *Daemon) *ControlService {
+	return &ControlService{daemon: d}
+}
+
+func (s *ControlService) DaemonStatus() map[string]*protocol.DriverPoolState {
+	out := make(map[string]*protocol.DriverPoolState, 0)
+	for language, pool := range s.daemon.Current() {
+		out[language] = pool.State()
+	}
+
+	return out
+}
+
+func (s *ControlService) InstancesStatus() ([]*protocol.DriverInstanceState, error) {
```

Comment:
I think naming can be improved here to make it more consistent. `Status`, `DaemonStatus`, `State`, `Instances`, `DriverInstanceState`, etc. are being mixed now. It could be unified to `Status`, `DaemonStatus`, `DriverInstanceStatus` and `DriverPoolStatus`, and remove the usage of `State`, etc.